src/expr/src/linear.rs

```diff
@@ -1642,6 +1642,11 @@ pub mod plan {
                     (Some(lower_bound), None) => {
                         (Some(Ok((row_builder.clone(), lower_bound, diff))), None)
                     }
+                    // NB: as of 2022-03-02 this code is dead, since
+                    // we already check if `lower_bound` is `None` above.
+                    // Thus, the actual behavior in this case
+                    // is *not* to emit only a retraction as the following code suggests,
+                    // but actually to emit nothing.
                     (None, Some(upper_bound)) => {
                         (None, Some(Ok((row_builder.clone(), upper_bound, -diff))))
```

Comment:
The fact that this does something non-trivial seems so misleading to me. I can see not wanting to change the type of `lower_build`, but do we not want to at least change this branch to `unreachable!`? (Or a `log::error` if we're worried about panics, I guess, but if this is a correctness issue, seems like a panic would be best.)

Reply:
A few points:
 
(1) With my original fix, #10976, it would be impossible for that to "flip back" without noticing, because the type system would prevent you from getting rid of the early return without adjusting other code (since the type of `lower_bound` would have changed).
(2) Right now, without this change, will _already_ have an issue if we get rid of the early return, because the dead code does a _different thing_. The live code emits nothing if there isn't a lower bound, whereas the dead code emits a retraction. Thus if we got rid of the early return, we would already have a bug as of now.

Given those two points, it seems to me like #10976 is strictly better than the present situation, and an `unreachable!` is at least no worse. However, since @frankmcsherry didn't seem open to modifying the code, I thought adding the comment would at least be better than the current situation, because it would help spare future readers the confusion I experienced when reading this code. But @antiguru and @benesch both didn't seem to want to add the comment, so I'm not sure how to proceed here.

Everyone, any thoughts on what we can do that will be acceptable to everyone?